shivyc/tree/expr_nodes.py

```diff
@@ -903,36 +904,67 @@ def __init__(self, expr):
     def make_il(self, il_code, symbol_table, c):
         """Make code for this node."""
         expr = self.expr.make_il(il_code, symbol_table, c)
-        if not expr.ctype.is_arith():
-            err = f"unary {self.descrip} requires arithmetic type operand"
+        if not self._check_type(expr):
+            err = f"{self.descrip} requires {self.opnd_descrip} type operand"
             raise CompilerError(err, self.expr.r)
+        # perform integer promotion
         if expr.ctype.size < 4:
             expr = set_type(expr, ctypes.integer, il_code)
         if self.cmd:
             out = ILValue(expr.ctype)
-            if expr.literal_val is not None:
-                val = -shift_into_range(expr.literal_val, expr.ctype)
+            # perform constant folding
+            if expr.literal_val:
```

Comment:
I think we actually want to keep `expr.literal_val is not None` here because otherwise if `expr_literal_val` is 0 then this condition does not trigger. 

That said, the implementation of literal values is kind of hacky right now. It would be better to create some kind of LiteralValue class in `il_gen.py` and inherit from that an `IntegerLiteralValue` class, `StringLiteralValue` class, etc. Then, we could have `expr.literal_val` be an instance of one of these classes if it has a literal value, or `None` if not. I'll have a busy the next few weeks, though, so unless you or someone else wants to implement that it may be a while until that improvement is made :)

Reply:
Thanks, Shivam. I changed that line, thinking it looked unnecessarily verbose, without considering the case where `expr._literal_val` is 0.

I'll have a look at improving the implementation of literal values based on your suggestions.

Reply:
Perfect, thank you! I've created an issue for this at #96 - please feel free to ask if you have any questions or want to run ideas by me.